packages/flutter/lib/src/painting/box_painter.dart

```diff
@@ -937,29 +938,42 @@ class BoxPainter {
     return _decoration.borderRadius > shortestSide ? shortestSide : _decoration.borderRadius;
   }
 
-  void _paintBackgroundColor(ui.Canvas canvas, Rect rect) {
-    if (_decoration.backgroundColor != null ||
-        _decoration.boxShadow != null ||
-        _decoration.gradient != null) {
-      switch (_decoration.shape) {
-        case Shape.circle:
-          assert(_decoration.borderRadius == null);
-          Point center = rect.center;
-          double radius = rect.shortestSide / 2.0;
-          canvas.drawCircle(center, radius, _backgroundPaint);
-          break;
-        case Shape.rectangle:
-          if (_decoration.borderRadius == null) {
-            canvas.drawRect(rect, _backgroundPaint);
-          } else {
-            double radius = _getEffectiveBorderRadius(rect);
-            canvas.drawRRect(new ui.RRect.fromRectXY(rect, radius, radius), _backgroundPaint);
-          }
-          break;
-      }
+  void _paintBox(ui.Canvas canvas, Rect rect, Paint paint) {
+    switch (_decoration.shape) {
+      case Shape.circle:
+        assert(_decoration.borderRadius == null);
+        Point center = rect.center;
+        double radius = rect.shortestSide / 2.0;
+        canvas.drawCircle(center, radius, paint);
+        break;
+      case Shape.rectangle:
+        if (_decoration.borderRadius == null) {
+          canvas.drawRect(rect, paint);
+        } else {
+          double radius = _getEffectiveBorderRadius(rect);
+          canvas.drawRRect(new ui.RRect.fromRectXY(rect, radius, radius), paint);
+        }
+        break;
     }
   }
 
+  void _paintShadows(ui.Canvas canvas, Rect rect) {
+    if (_decoration.boxShadow == null)
+      return;
+    for (BoxShadow boxShadow in _decoration.boxShadow) {
+      final Paint paint = new Paint()
+        ..color = boxShadow.color
+        ..maskFilter = new ui.MaskFilter.blur(ui.BlurStyle.normal, boxShadow._blurSigma);
```

Comment:
I tried a simple approach to caching Paint values. The cache grows and grows because of the interpolated offset, blurRadius etc values that come from animating the shadows. I think the same thing would occur if I just cached the MaskFilter values, since the floating point blur radius is part of the problem.

```
  static Map<BoxShadow, Paint> _boxShadowToPaintCache = new Map<BoxShadow, Paint>();
  Paint _boxShadowToPaint(BoxShadow boxShadow) {
    return _boxShadowToPaintCache.putIfAbsent(boxShadow, () {
      print("_boxShadowToPaintCache.length = ${_boxShadowToPaintCache.length} boxShadow=${boxShadow}");
      return new Paint()
        ..color = boxShadow.color
        ..maskFilter = new ui.MaskFilter.blur(ui.BlurStyle.normal, boxShadow._blurSigma);
    });
  }
```
I suppose we could avoid the cache misses by quantizing the boxShadow properties however it might be best to do some rasterization benchmarks first.